src/core/chunked_stream.js

```diff
@@ -267,24 +245,489 @@ var ChunkedStream = (function ChunkedStreamClosure() {
+        throw new MissingDataException(begin, end);
+      }
+    }
+
+    this.buffer.startOffset = 0;
+    this.buffer.start = begin;
+    this.buffer.end = end;
+  };
+
+  ChunkedStreamContinuous.prototype.onReceiveData = function
+      ChunkedStreamContinuous_onReceiveDataContinuous(begin, chunk) {
+    var chunkSize = this.chunkSize;
+    var data = new Uint8Array(chunk);
+
+    if (begin === 0 && data.byteLength % chunkSize !== 0) {
+      this.initialChunk = {
+        start:  0,
+        startOffset: 0,
+        end:    0,
+        buffer: this.buffer.buffer
+      };
+      this.buffer.buffer.set(data, this.initialChunk.end);
+      this.initialChunk.end += data.byteLength;
+
+      if (data.byteLength > chunkSize) {
+        data = data.subarray(0, data.byteLength -
+        data.byteLength % chunkSize);
+      } else {
+        return;
+      }
+    }
+
+    var end = begin + data.byteLength;
+    assert(begin % chunkSize === 0, 'Bad begin offset: ' + begin);
+
+    // Using this.length is inaccurate here since this.start can be moved
+    // See ChunkedStreamBase.moveStart()
+    var length = this.totalLength;
+    var beginChunk = Math.floor(begin / chunkSize);
+    var endChunk = Math.floor((end - 1) / chunkSize) + 1;
+
+    this.buffer.buffer.set(data, beginChunk * chunkSize);
+
+    assert(end % chunkSize === 0 || end === length,
+        'Bad end offset: ' + end);
+
+    var chunkEnd = chunkSize * beginChunk + data.byteLength;
+    if (this.loadedChunks[endChunk]) {
+      chunkEnd = this.loadedChunks[endChunk].end;
+    }
+
+    for (var curChunk = beginChunk; curChunk < endChunk; ++curChunk) {
+      if (!this.loadedChunks[curChunk]) {
+        this.loadedChunks[curChunk] = {
+          chunkOffset: chunkSize * curChunk,
+          start: 0,
+          startOffset: 0,
+          end:    chunkEnd,
+          data:  this.buffer.buffer
+        };
+        ++this.numChunksLoaded;
+      }
+    }
+    // Merge chunks to the left.
+    curChunk = beginChunk - 1;
+    while (curChunk >= 0 && this.loadedChunks[curChunk]) {
+      this.loadedChunks[curChunk--].end = chunkEnd;
+    }
+
+    if (this.numChunksLoaded === this.numChunks) {
+      // Stream is fully loaded, switch to fast getByte, prepareBuffer is no
+      // longer needed since memory is continuous.
+      this.initialChunk = null;
+      this.buffer.start = 0;
+      this.buffer.startOffset = 0;
+      this.buffer.end = this.buffer.buffer.byteLength;
+      this.getByte = this.getByteFastContinuous(this.buffer.buffer, this.end);
+      this.prepareBuffer = this.prepareBufferNop;
+    }
+  };
+
+  return ChunkedStreamContinuous;
+})();
+
+/* Fragmented version of ChunkedStream. Will allocate memory as needed.
+ Much more memory efficient for large PDF especially if not all pages need
+ to be loaded.
+ */
+var ChunkedStreamFragmented = (function ChunkedStreamFragmentedClosure() {
+
+    function ChunkedStreamFragmented(length, chunkSize, manager) {
```

Comment:
This line and the line below this line need to be indented with two spaces less.